exareme2/controller/services/flower/execution_engine.py

```diff
@@ -0,0 +1,122 @@
+from dataclasses import dataclass
+from typing import Dict
+from typing import List
+from typing import Optional
+
+from exareme2.controller import logger as ctrl_logger
+from exareme2.controller.services.flower.workers import GlobalWorker
+from exareme2.controller.services.flower.workers import LocalWorker
+
+
+@dataclass
+class Workers:
+    local_workers: List[LocalWorker]
+    global_worker: Optional[GlobalWorker] = None
+
+
+class AlgorithmExecutionEngine:
+    """
+    Manages distributed algorithm execution using the Flower framework,
+    accommodating configurations with varying numbers of local and potentially a global worker.
+    """
+
+    def __init__(self, request_id: str, workers: Workers):
+        self.logger = ctrl_logger.get_request_logger(request_id=request_id)
+        self.workers = workers
+
+    def start_flower(self, algorithm_name: str) -> Dict[str, Dict[str, int]]:
+        """
+        Initializes Flower servers and clients for a given algorithm.
+        """
+        if self.is_single_worker_mode():
+            return {
+                self.workers.local_workers[0].worker_id: self.start_on_single_worker(
+                    algorithm_name
+                )
+            }
+        else:
+            return self.start_on_multiple_workers(algorithm_name)
+
+    def stop_flower(self, process_ids: Dict[str, Dict[str, int]], algorithm_name: str):
+        """
+        Shuts down Flower servers and clients using their process IDs.
+        """
+        for worker_id, ids in process_ids.items():
+            self.stop_worker(worker_id, ids, algorithm_name)
+
+    def is_single_worker_mode(self) -> bool:
+        """
+        Checks if only one worker is available and it operates both as a server and a client.
+        """
+        return (
+            len(self.workers.local_workers) == 1 and self.workers.global_worker is None
+        )
+
+    def start_on_single_worker(self, algorithm_name: str) -> Dict[str, int]:
+        """
+        Starts both a server and a client on the single available worker.
+        """
+        worker = self.workers.local_workers[0]
+        return {
+            "server": worker.start_flower_server(
+                algorithm_name, len(self.workers.local_workers)
+            ),
+            "client": worker.start_flower_client(algorithm_name),
+        }
+
+    def start_on_multiple_workers(
+        self, algorithm_name: str
```

Comment:
We could initialize the execution engine with one node and the global node being the same local node. 

That way we could skip the logic of "signle_node_execution.

WDYT?

Reply:
You mean  like {"globalworker": "same_id", "localworker":"same_id"}.
If you mean that I think it kinda hides the logic I am not sure that it will be more clear